api/include/opentelemetry/common/attribute_value.h

```diff
@@ -10,20 +10,30 @@
 OPENTELEMETRY_BEGIN_NAMESPACE
 namespace common
 {
-using AttributeValue = nostd::variant<bool,
-                                      int,
-                                      int64_t,
-                                      unsigned int,
-                                      uint64_t,
-                                      double,
-                                      nostd::string_view,
-                                      nostd::span<const bool>,
-                                      nostd::span<const int>,
-                                      nostd::span<const int64_t>,
-                                      nostd::span<const unsigned int>,
-                                      nostd::span<const uint64_t>,
-                                      nostd::span<const double>,
-                                      nostd::span<const nostd::string_view>>;
+using AttributeValue =
+    nostd::variant<bool,
+                   int32_t,
+                   int64_t,
+                   uint32_t,
+                   uint64_t,
+                   double,
+                   nostd::string_view,
+#ifdef HAVE_CSTRING_TYPE
```

Comment:
QQ For my sanity:  Do you know if adding types like this breaks ABI compatibility?

IIRC:  std::variant assigns a unique integer to each type, and this integer is defined by its memory location at load (at least in ABSL this was true).   Theoretically, this means we can add a type without breaking source compatibility but AFAIK this would definitely break ABI compatibility?

Given new ABI requirements from the spec, we may need to comment on that in the README.

Reply:
It's a good question. My hope is that a change like this does not create ABI compat issue. I can verify this.

For two types I'm adding:

- adding `const char *` support is a non-breaking change that follows spec, that we should add before `v1.0.0`. Even protobuf / gRPC code has construct that respects string input from either `string` or `string_view` or `const char *`. I hope we can just remove the "feature gate" from that soon, in a follow-up PR.

- adding byte buffers (not spec'd, but I filed a request for spec) - I anticipate that byte buffers would eventually be added to spec. And, for sure, we will needed this for our customer our flows in Azure.

I'd like to keep some flexibility here, for the case where we statically compile the SDK into product that would need both `const char *` and `uint8_t[]` support. Feature flags are off by default, and are not affecting the mainline OpenTelemetry builds. Unfortunately there is no better way to add types to SDK at this point.

Reply:
I'll keep it under `#ifdef` - NOT enabled by default, and log a separate issue to investigate if adding new types to variant would break ABI compatibility.

Reply:
https://github.com/open-telemetry/opentelemetry-cpp/issues/474